### PR TITLE
crate_universe: Support fetching crates with git branch, tag or rev

### PR DESCRIPTION
**Pull request by ted-logan**
In crate_universe, allow crate.spec() to be used with a git branch or tag, in addition to a rev.

This picks up an earlier open pull request (https://github.com/bazelbuild/rules_rust/pull/1283), resolves a merge conflict in the generated doc, and adds test cases to check that several valid combinations of crate specifications can be serialized and deserialized.

### CODE REVIEW

File: crate_universe/private/crate.bzl [hunk 1]
```diff
@@ -34,7 +36,9 @@ def _spec(
         default_features (bool, optional): Maps to the `default-features` flag.
         features (list, optional): A list of features to use for the crate
         git (str, optional): The Git url to use for the crate. Cannot be used with `version`.
-        rev (str, optional): The git revision of the remote crate. Tied with the `git` param.
+        branch (str, optional): The git branch of the remote crate. Tied with the `git` param. Only one of branch, tag or rev may be specified.
```
Comment (UebelAndre) [inline crate_universe/private/crate.bzl:39]: Can you update the docs for `branch` and `tag` to say something about `rev` being the recommended value?

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Thanks! One request 😄

File: crate_universe/private/crate.bzl (hunk 1 above)
Comment (ted-logan) [reply crate_universe/private/crate.bzl]: Done.

**Comment by UebelAndre** [comment]
Sorry, merging a ton of stuff from the weekend. Can you rebase once more?

**Comment by UebelAndre** [comment]
hehe, one more rebase. Normally there's a button I can click to rebase for you but you need to allow maintainers to edit your commits. All PRs that were in the queue to merge are merged though so this can be next 😅

**Review by UebelAndre — APPROVED** [review]

**Pull request merged**